### DISCUSSION TITLE
How to retrieve the user agent in Gatsby-SSR.js for use with React-Responsive?

### DISCUSSION THREAD

**Discussion by lukasdcns** [gatsbyjs/gatsby#37763] [Help]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

I am currently facing an issue with integrating react-responsive into my Gatsby (V4) application. Specifically, I am having trouble with server-side rendering (SSR) and how to retrieve the user agent in `gatsby-ssr.js`.

My goal is to return a different context in `gatsby-ssr.js` based on the type of device, and I need to retrieve the user agent to accomplish this.

As a relatively new Gatsby developer, I am struggling to find a solution to retrieve the user agent in gatsby-ssr.js. Can you please provide some guidance on how to achieve this?

Thank you in advance for your assistance.

### Reproduction Link

...

### Steps to Reproduce

1.
2.
3.
...

### Expected Result

```
if (desktop) {
    <ResponsiveContext.Provider value={{ width: 1440 }}>
        {element}
    </ResponsiveContext.Provider>
} else {
    <ResponsiveContext.Provider value={{ width: 991 }}>
        {element}
    </ResponsiveContext.Provider>
}
```

### Actual Result

...

### Environment

```shell
System:
    OS: macOS 13.2.1
    CPU: (8) x64 Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.12.1 - /usr/local/bin/node
    npm: 8.19.2 - /usr/local/bin/npm
  Browsers:
    Chrome: 111.0.5563.64
    Firefox: 107.0
    Safari: 16.3
  npmPackages:
    gatsby: ^4.22.0 => 4.22.0 
    gatsby-cli: ^4.24.0 => 4.24.0 
    gatsby-plugin-algolia: ^0.26.0 => 0.26.0 
    gatsby-plugin-google-tagmanager: ^4.17.0 => 4.19.0 
    gatsby-plugin-image: ^2.17.0 => 2.19.0 
    gatsby-plugin-less: ^6.17.0 => 6.19.0 
    gatsby-plugin-manifest: ^4.17.0 => 4.19.0 
    gatsby-plugin-offline: ^5.17.0 => 5.19.0 
    gatsby-plugin-preact: ^6.17.0 => 6.19.0 
    gatsby-plugin-prismic-previews: ^5.2.8 => 5.2.8 
    gatsby-plugin-react-helmet: ^5.17.0 => 5.19.0 
    gatsby-plugin-robots-txt: ^1.7.1 => 1.7.1 
    gatsby-plugin-sass: ^5.17.0 => 5.19.0 
    gatsby-plugin-sharp: ^4.17.0 => 4.19.0 
    gatsby-plugin-sitemap: ^5.17.0 => 5.19.0 
    gatsby-plugin-webpack-bundle-analyser-v2: ^1.1.22 => 1.1.27 
    gatsby-remark-images: ^6.17.0 => 6.19.0 
    gatsby-remark-lazy-load: ^1.0.2 => 1.0.2 
    gatsby-source-filesystem: ^4.17.0 => 4.19.0 
    gatsby-source-prismic: ^5.2.9 => 5.2.9 
    gatsby-transformer-remark: ^5.17.0 => 5.19.0 
    gatsby-transformer-sharp: ^4.17.0 => 4.19.0 
  npmGlobalPackages:
    gatsby-cli: 5.7.0
```

**Comment by LekoArts** [comment]
The `gatsby-ssr` file is run during SSR (hence the name) and therefore you can't get a user agent there. Use [`gatsby-browser`](https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/) to access that information.

I'd personally refrain from using react-responsive and use pure CSS media queries. Less headaches about handling the default case during SSR and rehydration errors.
